src/main/java/com/hivemq/persistence/RetainedMessage.java

```diff
@@ -85,6 +114,50 @@ public RetainedMessage(@NotNull final PUBLISH publish, @Nullable final Long payl
         this.timestamp = publish.getTimestamp();
     }
 
+    public RetainedMessage copyWithoutPayload() {
+        return new RetainedMessage(
+                null,
+                qos,
+                payloadId,
+                messageExpiryInterval,
+                userProperties,
+                responseTopic,
+                contentType,
+                correlationData,
+                payloadFormatIndicator,
+                timestamp);
+    }
+
+    public int getEstimatedSizeInMemory() {
+
+        if (sizeInMemory != SIZE_NOT_CALCULATED) {
+            return sizeInMemory;
+        }
+        int size = 0;
+        // The payload size is not calculated because the payload is removed before the message is stored
+        size += ObjectMemoryEstimation.enumSize(); // QoS
+        size += ObjectMemoryEstimation.longWrapperSize(); // Payload ID
+        size += ObjectMemoryEstimation.longSize(); // expiry interval
+
+        size += 24; //User Properties Overhead
+        for (final MqttUserProperty userProperty : getUserProperties().asList()) {
+            size += 12; //UserProperty Object Overhead
```

Comment:
now the fields are missing, the overhead should probably be 24 (12 object overhead + 4 ref for name + 4 ref for value + padding)